packages/backend/src/open_payments/payment/incoming/service.ts

```diff
@@ -278,6 +329,50 @@ async function getWalletAddressPage(
   })
 }
 
+async function approveIncomingPayment(
+  deps: ServiceDependencies,
+  id: string
+): Promise<IncomingPayment | IncomingPaymentError> {
+  return deps.knex.transaction(async (trx) => {
+    const payment = await IncomingPayment.query(trx)
+      .findById(id)
+      .forUpdate()
+      .withGraphFetched('[asset, walletAddress]')
+
+    if (!payment) return IncomingPaymentError.UnknownPayment
+    if (payment.state !== IncomingPaymentState.Pending)
+      return IncomingPaymentError.WrongState
+
+    await payment.$query(trx).patch({
+      approvedAt: new Date(Date.now())
+    })
+
+    return await addReceivedAmount(deps, payment)
+  })
+}
```

Comment:
I can cancel a pending incoming payment and then approve an incoming payment and both timestamps will be set.

How do we want to handle that? I think its probably fine to switch from canceled to approved but it should unset the other one. This applies to cancelIncomingPayment as well. I can approve then cancel a payment and both timestamps are set.

Tangential, and probably not a problem, but I can approve over and over again and it resets the `approvedAt`. I think this might be the best behavior though. We could instead just return the payment without patching (not erroring) but I feel like the way we have it now is more predictable... unless we have some reason not to lose the original `approvedAt` time but I dont think so. (note that this applies to cancelledAt too).

Reply:
Wow, great catch. Don't know how I missed this. Instead of checking timestamps, I would suggest that when IP is approved/cancelled, that we check if opposite field is already set. If it is, then throw error, otherwise, check if it was already approved/cancelled and don't update timestamp

```
// approve IP
incomingPayment := getIncomingPayment()
if incomingPayment.cancelledAt then
    throw Error('IP was already cancelled')
elif incomingPayment.approvedAt then
    return incomingPayment
end

incomingPayment.approvedAt = now()

return incomingPayment
```

Something like this

Reply:
I added additional checks when approve or cancel incoming payment method is called. I also added tests to checks exactly what you wrote above.